### PR TITLE
Implement CognitoIdentityManager

### PR DESCRIPTION
**Pull request by gshaw1997**
This PR adds the CognitoIdentityManager module. This module currently retrieves an `OpenIdToken` for a given user for use in receiving credentials from AWS's Cognito service to access resources on AWS.

In its current state, this module will be used to retrieve credentials for uploading files to S3.

**Update**
This PR adds the `openId` token to the response payload of the login route (POST `/users/tokens`)

### CODE REVIEW

File: src/CognitoIdentityManager/adapters/ExpressRouterAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Builds the Express Router for this module
+ *
+ * @export
+ * @returns
+ */
+export function buildRouter() {
+  const router = Router();
+  router.get('/users/:userId/open-id-token', getOpenIdToken);
```
Comment (DonSeannelly) [inline src/CognitoIdentityManager/adapters/ExpressRouterAdapter.ts]: What about adding this to our existing `tokens` endpoint, so that we can just specify `.../tokens?openIdToken` or `.../tokens?bearer`? That way we leave our route structures the same, and now the tokens resource can support multiple types of tokens.

I think this would be especially helpful since there's a good chance we might want to implement OAuth in the future.

TL;DR Is an OpenID Token its own resource, or is it a type of resource?

Thoughts?

Comment (gshaw1997) [reply src/CognitoIdentityManager/adapters/ExpressRouterAdapter.ts]: I believe that the `OpenIdToken` is a different resource since there is no commonality between the resources' interfaces. 

From my understanding, its best practice to reserve query parameters for filtering, sorting, paging, and field selection to allow a change in shape of the response interface.

Comment (DonSeannelly) [reply src/CognitoIdentityManager/adapters/ExpressRouterAdapter.ts]: > I believe that the `OpenIdToken` is a different resource since there is no commonality between the resources' interfaces.

Can you elaborate on this?


I'm seeing it as the query parameter is in fact a filter applied to a user's tokens. So the base resource would include all tokens for a user (leaving the route better open to extension), and query parameters could be applied to retrieve just a subset.

Whiteboard on Apr 19, 2019 at 9_20_13 AM

Comment (gshaw1997) [reply src/CognitoIdentityManager/adapters/ExpressRouterAdapter.ts]: Thank you for the visualization; that clears things up. I initially misunderstood your proposal. I thought that you were suggesting that we change the response type based on the query parameter.

**Example**

File: src/drivers/AuthRouteHandler.ts [hunk 2]
```diff
@@ -111,6 +114,39 @@ export default class AuthRouteHandler {
...
+            tokens.openId = await CognitoIdentityManager.adapter.getOpenIdToken(
+              {
+                requester,
+                userId
+              }
+            );
+          }
+        } else {
+          // Grab all token properties
+          tokens.openId = await CognitoIdentityManager.adapter.getOpenIdToken({
```
Comment (DonSeannelly) [inline src/drivers/AuthRouteHandler.ts]: I'm not following the if-else here. It looks like the same code is run in both blocks of the conditional?

Comment (gshaw1997) [reply src/drivers/AuthRouteHandler.ts]: The if case provides a way to select specific properties of the token object as suggested earlier within this PR. If no selection is performed, all properties will be returned to the token object. Right now there is only one property `openId`.

File: src/CognitoIdentityManager/Interactor.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+AWS.config.credentials = SDK_CONFIG.credentials;
+AWS.config.region = SDK_CONFIG.region;
+
+const Cognito = new AWS.CognitoIdentity();
+const DEVELOPER_PROVIDER = process.env.COGNITO_DEVELOPER_PROVIDER;
+const IDENTITY_POOL_ID = process.env.COGNITO_IDENTITY_POOL_ID;
+const ADMIN_IDENTITY_POOL_ID = process.env.COGNITO_ADMIN_IDENTITY_POOL_ID;
+
+const TOKEN_DURATION = 86400;
```
Comment (DonSeannelly) [inline src/CognitoIdentityManager/Interactor.ts]: I think a comment here that specifies units of measurement would help developers understand this better

File: src/interactors/AuthenticationInteractor.ts [hunk 4]
```diff
@@ -22,38 +25,40 @@ export async function login(
...
-        emailVerified: user.emailVerified,
-        accessGroups: user.accessGroups
-      };
-      return {token, user: userResponse};
+    const authenticated = await hasher.verify(password, user.password);
+    if (!authenticated) {
+      throw invalidCredentialsError;
     }
-    return authenticated;
+    const bearer = TokenManager.generateToken(user);
```
Comment (DonSeannelly) [inline src/interactors/AuthenticationInteractor.ts]: We might consider renaming this to `generateBearerToken` now that we are working with multiple types.

File: src/drivers/AuthRouteHandler.ts [hunk 5]
```diff
@@ -96,33 +99,36 @@ export default class AuthRouteHandler {
...
-          responder.sendUser(req.user);
-        } catch (e) {
-          responder.sendOperationError('Invalid token');
-        }
-      });
+    router.route('/users/tokens').get(async (req, res) => {
+      const responder = this.responseFactory.buildResponder(res);
+      try {
+        const requester: UserToken = req.user;
+        const token = await AuthInteractor.refreshToken({
```
Comment (DonSeannelly) [inline src/drivers/AuthRouteHandler.ts]: As per our conversation, we don't want to refresh the token here due to the large volume of requests we have hitting the route during the tokens normal activation window. Instead, we should store the OpenID token in the client.
